Lib/test/test_ntpath.py

```diff
@@ -107,23 +107,23 @@ def test_splitdrive(self):
         tester('ntpath.splitdrive("//conky/mountpoint/foo/bar")',
                ('//conky/mountpoint', '/foo/bar'))
         tester('ntpath.splitdrive("\\\\\\conky\\mountpoint\\foo\\bar")',
-            ('', '\\\\\\conky\\mountpoint\\foo\\bar'))
+            ('\\\\\\conky', '\\mountpoint\\foo\\bar'))
         tester('ntpath.splitdrive("///conky/mountpoint/foo/bar")',
-            ('', '///conky/mountpoint/foo/bar'))
+            ('///conky', '/mountpoint/foo/bar'))
         tester('ntpath.splitdrive("\\\\conky\\\\mountpoint\\foo\\bar")',
-               ('', '\\\\conky\\\\mountpoint\\foo\\bar'))
+               ('\\\\conky', '\\\\mountpoint\\foo\\bar'))
         tester('ntpath.splitdrive("//conky//mountpoint/foo/bar")',
-               ('', '//conky//mountpoint/foo/bar'))
+               ('//conky', '//mountpoint/foo/bar'))
         # Issue #19911: UNC part containing U+0130
         self.assertEqual(ntpath.splitdrive('//conky/MOUNTPOİNT/foo/bar'),
                          ('//conky/MOUNTPOİNT', '/foo/bar'))
         # gh-81790: support device namespace, including UNC drives.
         tester('ntpath.splitdrive("//?/c:")', ("//?/c:", ""))
         tester('ntpath.splitdrive("//?/c:/")', ("//?/c:", "/"))
         tester('ntpath.splitdrive("//?/c:/dir")', ("//?/c:", "/dir"))
-        tester('ntpath.splitdrive("//?/UNC")', ("", "//?/UNC"))
-        tester('ntpath.splitdrive("//?/UNC/")', ("", "//?/UNC/"))
-        tester('ntpath.splitdrive("//?/UNC/server/")', ("//?/UNC/server/", ""))
+        tester('ntpath.splitdrive("//?/UNC")', ("//?/UNC", ""))
+        tester('ntpath.splitdrive("//?/UNC/")', ("//?/UNC/", ""))
+        tester('ntpath.splitdrive("//?/UNC/server/")', ("//?/UNC/server", "/"))
```

Comment:
This seems like a change we don't want (and the other instances below). I know you've fixed our own isabs implementation to not be caught out here, but I bet others have done it (perhaps accidentally/implicitly).

The result isn't meaningful either way, but this is a "close enough" result that we're breaking, and I think I'd rather not (but might be wrong... it's late)

Reply:
@eryksun suggested this change here, for conformity with `nt._path_splitroot()`: https://github.com/python/cpython/pull/100351#discussion_r1057871700

I've undone it for now. Eryk, what do you think?

Reply:
The intention of my suggestion was to be strictly compatible with how the API handles these cases. For example:

```pycon
>>> nt._path_splitroot('//server//spam')
('//server/', '/spam')
>>> nt._getfullpathname('//server//spam')
'\\\\server\\spam'
```

It's not that I like this result. It's dubious to split the root before an empty share because an invalid path can be normalized as a valid path. Python's builtin `nt._path_normpath()` function preserves the empty share component:

```pycon
>>> nt._path_normpath('//server//spam')
'\\\\server\\\\spam'
```

I know you'd like to add an `os.path.splitroot()` function for pathlib to use. For the C implementation, `nt._path_splitoot()`, we can either modify the current implementation that's based on WinAPI `PathCchSkipRoot()`, and also use `PathCchSkipRoot()` in `nt._path_normpath()` for consistency, or we can develop our own builtin function that matches the behavior of `_path_normpath`, presuming it's fixed to support "//?/UNC/" paths.

---

`isabs()` has always incorrectly returned false for some device and UNC paths, such as "//./C:", "//server/share", and "//server/". Such paths never depend on the current working directory. I don't see a downside to fixing this. I think even developers with no experience on Windows would guess that these are absolute paths.

`isabs()` has always incorrectly returned true for rooted paths such as "/spam". This mistake is entrenched and repeated all over the place. Most programmers simply assume that a rooted path is absolute. I don't think we can fix this in the public `isabs()` function. However, we should have a correct implementation for internal use where it matters. For example, due to this mistake, `realpath()` misbehaves when resolving a dangling symlink that targets a rooted path. For example:

```pycon
>>> os.symlink(r'\Windows\missing', 'link')
>>> os.path.abspath('link')
'C:\\Temp\\link'
>>> os.path.realpath('link')
'\\Windows\\missing'
```
If `isabs()` returned false for "\\Windows\\missing", as it should, then the target path of the symlink would be resolved against its opened path to yield the correct result "C:\\Windows\\missing".

Reply:
> I know you'd like to add an `os.path.splitroot()` function for pathlib to use.

Yes, though I'm only planning to add a Python implementation - a modification of `splitdrive()`. If someone wants to add a version that uses `nt._path_splitroot()`, I'll leave it to them to worry about maintaining bug-for-bug compatibility between the two implementations. It seems premature to worry here.

Reply:
(Though I of course appreciate you pointing this out, and it's possible I will be that "someone" from my previous comment)

Reply:
As much as I want the native implementation to be just the OS API and nothing more, it seems like that isn't going to be viable here. We need (at some point) to handle `\\?\UNC` even if the OS doesn't.

For invalid/impossible paths, I'm less concerned about consistency between the native and Python implementations. Round-tripping matters still, but if slashes end up on different sides of the result that's not a big deal. (Not changing the result between versions is more important.)

> Such paths never depend on the current working directory. I don't see a downside to fixing this.

Me neither. That's a good definition for `isabs`, and it definitely means we should fix the ones that start with `\`.

Being able to have equivalent `isabs` results on valid paths using both Python and native implementations would be ideal.

Reply:
When Steve added `nt._path_splitroot()` there were too many differences to adopt it as the basis of `splitdrive()` and `normpath()`. IIRC, there were even more differences with the older `PathSkipRootW()` function that would have been needed for Python 3.8 on Windows 7.

Using `PathCchSkipRoot()` has remained in the back of my mind. You already modified `splitdrive()` to support "\\\\?\\UNC" share drives vs "\\\\.\\UNC" device drives, just like `PathCchSkipRoot()`. With this proposed change, `splitdrive()` will also support the invalid UNC paths that `PathCchSkipRoot()` supports. At the very least we're getting closer to the API, even if we end up implementing `nt._path_splitroot()` from scratch.